src/unsafe_string.rs

```diff
@@ -52,48 +54,113 @@ trait HeaderMut<'a>: ThinMutExt<'a, Header> {
+            *s_c = HashSet::new();
+        }
     }
 }
 
-pub(crate) fn reinit_cache() {
-    unsafe {
-        STRING_CACHE = Some(HashSet::new());
+pub(crate) fn init_cache(thread_safe: bool) -> Result<(), String> {
+    let s_c = unsafe { &*addr_of_mut!(STRING_CACHE) };
+    s_c.set(if thread_safe {
+        StringCache::ThreadSafe(Mutex::new(HashSet::new()))
+    } else {
+        StringCache::ThreadUnsafe(HashSet::new())
+    })
+    .map_err(|_| "Cache is already initialized".to_owned())
+}
+
+fn get_cache() -> &'static mut StringCache {
+    let s_c = unsafe { &mut *addr_of_mut!(STRING_CACHE) };
+    s_c.get_or_init(|| StringCache::ThreadUnsafe(HashSet::new()));
+    s_c.get_mut().unwrap()
+}
+
+fn is_thread_safe() -> bool {
+    match get_cache() {
+        StringCache::ThreadSafe(_) => true,
+        StringCache::ThreadUnsafe(_) => false,
     }
 }
 
-#[cfg(not(feature = "ctor"))]
-unsafe fn get_cache() -> &'static mut HashSet<WeakIString> {
-    init_cache();
-    STRING_CACHE.as_mut().unwrap()
+enum CacheGuard {
+    ThreadUnsafe(&'static mut HashSet<WeakIString>),
+    ThreadSafe(MutexGuard<'static, HashSet<WeakIString>>),
 }
 
-#[cfg(feature = "ctor")]
-unsafe fn get_cache() -> &'static mut HashSet<WeakIString> {
-    STRING_CACHE.as_mut().unwrap()
+impl CacheGuard {
+    fn get_or_insert<'a>(
+        &mut self,
+        value: &str,
+        f: Box<dyn FnOnce(&str) -> WeakIString + 'a>,
+    ) -> &WeakIString {
+        match self {
+            CacheGuard::ThreadSafe(c_g) => c_g.get_or_insert_with(value, |val| f(val)),
+            CacheGuard::ThreadUnsafe(c_g) => c_g.get_or_insert_with(value, |val| f(val)),
+        }
+    }
+
+    fn get_val(&self, val: &str) -> Option<&WeakIString> {
+        match self {
+            CacheGuard::ThreadSafe(c_g) => c_g.get(val),
+            CacheGuard::ThreadUnsafe(c_g) => c_g.get(val),
+        }
+    }
+
+    fn remove_val(&mut self, val: &str) -> bool {
+        match self {
+            CacheGuard::ThreadSafe(c_g) => c_g.remove(val),
+            CacheGuard::ThreadUnsafe(c_g) => c_g.remove(val),
+        }
+    }
+
+    #[cfg(test)]
+    fn check_if_empty(&self) -> bool {
+        match self {
+            CacheGuard::ThreadSafe(c_g) => c_g.is_empty(),
+            CacheGuard::ThreadUnsafe(c_g) => c_g.is_empty(),
+        }
+    }
+
+    #[cfg(test)]
+    fn shrink(&mut self) {
+        match self {
+            CacheGuard::ThreadSafe(c_g) => c_g.shrink_to_fit(),
+            CacheGuard::ThreadUnsafe(c_g) => c_g.shrink_to_fit(),
+        }
+    }
+}
+
+fn get_cache_guard() -> CacheGuard {
+    let s_c = get_cache();
+    match s_c {
+        StringCache::ThreadUnsafe(s_c) => CacheGuard::ThreadUnsafe(s_c),
+        StringCache::ThreadSafe(s_c) => {
+            CacheGuard::ThreadSafe(s_c.lock().expect("Mutex lock should succeed"))
+        }
+    }
 }
 
 struct WeakIString {
     ptr: NonNull<Header>,
 }
 
+unsafe impl Sync for WeakIString {}
+unsafe impl Send for WeakIString {}
```

Comment:
Do we need it? Do the tests, if you write them, show that the type does not get the blanket implementations of these? If the type has no mutable methods and some other easy requirements (probably, should also be a Copy), IIRC, it should be Sync and Send. Don't know, can't remember, I usually always have tests which confirm that the certain type implements Send and Sync.

Reply:
Actually, we do not need it anymore, let me remove it.